typescript/agentkit/src/action-providers/flaunch/flaunchActionProvider.ts

```diff
@@ -0,0 +1,638 @@
+            path: [
+              {
+                fee: 0,
+                tickSpacing: 60,
+                hookData: "0x",
+                hooks: FLETHHooksAddress[chainId],
+                intermediateCurrency: FLETHAddress[chainId],
+              },
+              {
+                fee: 0,
+                tickSpacing: 60,
+                hooks: FlaunchPositionManagerAddress[chainId],
+                hookData: "0x",
+                intermediateCurrency: args.coinAddress,
+              },
+            ],
+          },
+        ],
+      });
+      amountOutMin = getAmountWithSlippage(
+        quoteResult.result[0], // amountOut
+        (args.slippagePercent / 100).toFixed(18).toString(),
+        swapType,
+      );
+
+      const { commands, inputs } = ethToMemecoin({
+        sender: walletProvider.getAddress() as Address,
+        memecoin: args.coinAddress as Address,
+        chainId: Number(chainId),
+        referrer: zeroAddress,
+        swapType: swapType!,
+        amountIn: amountIn,
+        amountOutMin: amountOutMin,
+      });
+
+      const data = encodeFunctionData({
+        abi: UNIVERSAL_ROUTER_ABI,
+        functionName: "execute",
+        args: [commands, inputs],
+      });
+
+      const hash = await walletProvider.sendTransaction({
+        to: UniversalRouterAddress[chainId],
+        data,
+        value: amountIn,
+      });
+
+      const receipt = await walletProvider.waitForTransactionReceipt(hash);
+      const swapAmounts = getSwapAmountsFromReceipt({
+        receipt,
+        coinAddress: args.coinAddress as Address,
+        chainId: Number(chainId),
+      }) as BuySwapAmounts;
+
+      const coinSymbol = await walletProvider.readContract({
+        address: args.coinAddress as Address,
+        abi: ERC20_ABI,
+        functionName: "symbol",
+      });
+
+      return `Bought ${formatEther(swapAmounts.coinsBought)} $${coinSymbol} for ${formatEther(swapAmounts.ethSold)} ETH\n
+        Tx hash: [${hash}](${NETWORK_ID_TO_VIEM_CHAIN[networkId].blockExplorers?.default.url}/tx/${hash})`;
+    } catch (error) {
+      return `Error buying coin: ${error}`;
+    }
+  }
+
+  /**
+   * Buys a flaunch coin using Coin input.
+   *
+   * @param walletProvider - The wallet provider instance for blockchain interactions
+   * @param args - Arguments defined by BuyCoinSchema
+   * @returns A promise that resolves to a string describing the transaction result
+   */
+  @CreateAction({
+    name: "buyCoinWithCoinInput",
+    description: `
+This tool allows buying a flaunch coin using ETH, when the user has specified the Coin amount to buy.
+
+It takes:
+- coinAddress: The address of the flaunch coin to buy
+- amountOut: The quantity of the flaunch coin to buy, in whole units
+  Examples:
+  - 1000 coins
+  - 1_000_000 coins
+- slippagePercent: (optional) The slippage percentage. Default to 5%
+    `,
+    schema: BuyCoinWithCoinInputSchema,
+  })
+  async buyCoinWithCoinInput(
```

Comment:
`buyCoinWithETHInput` and `buyCoinWithCoinInput` look offly similar. Please extract the core work to a private function that takes the specific requirements needed to fulfill the task, and then have these two public functions wrap it and resolving the necessary variables as needed